platform/pulp/plugin/changeset/main.py

```diff
@@ -0,0 +1,200 @@
+from collections import Sized, Iterable
+from gettext import gettext as _
+from logging import getLogger
+
+from django.db.utils import IntegrityError
+from django.db import transaction
+
+from pulp.app.models import ProgressBar, DownloadCatalog, RepositoryContent
+from pulp.download import Batch
+
+from ..tasking import Task
+
+from .iterator import DownloadIterator
+from .report import AddReport, DeleteReport
+
+
+log = getLogger(__name__)
+
+
+class ChangeSet:
+    """
+    A set of changes to be applied to a repository.
+
+    Attributes:
+        importer (pulp.plugin.Importer): An importer.
+        adds (SizedIterable): The content to be added to the repository.
+        deletes (SizedIterable): The content IDs to be removed.
+        deferred (bool): Downloading is deferred.  When true, downloads are not performed.
+
+    Examples:
+
+            >>>
+            >>> # TODO: Add example.
+            >>>
+    """
+
+    def __init__(self, importer, adds=(), deletes=()):
+        """
+        Args:
+            importer (pulp.plugin.Importer): An importer.
+            adds (SizedIterable): The content to be added to the repository.
+            deletes (SizedIterable): The content IDs to be removed.
+
+        Notes:
+            The content to be added may already exist but not be associated
+            to the repository. Existing content is fetched and used instead
+            of provided content models as needed.
+        """
+        self.importer = importer
+        self.adds = adds
+        self.deletes = deletes
+        self.deferred = False
+
+    def _associate(self, content):
+        """
+        Associate content.
+
+        Args:
+            content (RemoteContent): The content to be associated.
+        """
+        log.debug(_('Associate content.'))
+        try:
+            association = RepositoryContent(
+                repository=self.importer.repository,
+                content=content)
+            association.save()
+        except IntegrityError:
+            # Duplicate
+            pass
+
+    def _disassociate(self, content):
```

Comment:
I recall discussion of disassociate vs unassociate. I don't recall how that ended. Should this all be consistent at least?